### PR TITLE
Install Checkov with docker/pip3/pipenv

### PR DESCRIPTION
**Pull request by elaygl**
**In This PR**
- Try to install Checkov with docker/pip3/pipenv
- Checkov runner integrate with a suitable Checkov executable

- [X] I've reviewed my own code

### CODE REVIEW

File: src/checkovInstaller.ts [hunk 1]
```diff
@@ -1,25 +1,48 @@
...
 import { Logger } from 'winston';
 import { asyncExec } from './utils';
 
-const updateCheckovWithPip3 = async (logger: Logger): Promise<void> => {
+const installOrUpdateCheckovWithPip3 = async (logger: Logger): Promise<string> => {
     try {
         logger.info('Trying to install Checkov using pip3.');
-        await asyncExec('pip3 install -U checkov -i https://pypi.org/simple/');
-        logger.info('Checkov installed successfully using pip3.');
+        await asyncExec('pip3 install -U --user checkov -i https://pypi.org/simple/');
```
Comment (schosterbarak) [inline src/checkovInstaller.ts:8]: you can add --verbose if you want

File: src/checkovInstaller.ts [hunk 2]
```diff
@@ -1,25 +1,48 @@
...
 export interface CheckovInstallation {
-    checkovPython: CheckovPython;
+    checkovInstallationMethod: CheckovInstallationMethod;
+    checkovPath: string;
 }
 
 export const installOrUpdateCheckov = async (logger: Logger): Promise<CheckovInstallation> => {
+    const checkovInstalledBeforePlugin = await isCheckovInstalledBeforePlugin();
+    if (checkovInstalledBeforePlugin) {
+        logger.info('Checkov is installed on machine before plugin.');
```
Comment (schosterbarak) [inline src/checkovInstaller.ts:37]: but what if it's not from the right version? users would not receive the value? how do you know it's not working for existing checkov users?

File: src/checkovInstaller.ts [hunk 3]
```diff
@@ -1,25 +1,75 @@
...
         logger.error('Failed to install or update Checkov using pip3. Error:', { error });
-        throw new Error('Failed to install or update Checkov using pip3');
+        return null;
+    }
+};
+
+const installOrUpdateCheckovWithPipenv = async (logger: Logger): Promise<string | null> => {
+    try {
+        logger.info('Trying to install Checkov using pipenv.');
+        await asyncExec('pipenv --python 3 install checkov');
```
Comment (SteveVaknin) [inline src/checkovInstaller.ts:37]: Did you check it works? 
We tried to add it before and had a permission issue if I remember correctly (Something with a pip file).

Comment (elaygl) [reply src/checkovInstaller.ts:37]: It worked for me. But it does create a Pipfile in the project dir.
I'm working on a fix for that now.

File: src/extension.ts [hunk 4]
```diff
@@ -134,7 +135,13 @@ export function activate(context: vscode.ExtensionContext): void {
         try {
             setSyncingStatusBarItem();
             const filePath = fileUri ? fileUri.fsPath : editor.document.fileName;
-            const checkovResponse = await runCheckovScan(logger, extensionVersion, filePath, token, cancelToken);
+            
+            if (!checkovInstallation) {
```
Comment (SteveVaknin) [inline src/extension.ts:139]: We already have the `extensionReady` flag that indicates checkov has installed correctly, why do we need this variable too?
If you want to check also here that checkov is installed, use the existing variable.

File: src/extension.ts [hunk 5]
```diff
@@ -134,7 +135,13 @@ export function activate(context: vscode.ExtensionContext): void {
...
+                logger.error('Checkov is not installed.');
```
Comment (SteveVaknin) [inline src/extension.ts:140]: Add consequence, like `Checkov is not installed, aborting scan.`

**Review by SteveVaknin — APPROVED** [review]
Looks good, minor comments

**Pull request merged**